src/coreclr/jit/optcse.cpp

```diff
@@ -3684,6 +3684,14 @@ bool Compiler::optIsCSEcandidate(GenTree* tree)
         case GT_RETURN:
             return false; // Currently the only special nodes that we hit
                           // that we know that we don't want to CSE
+        case GT_NULLCHECK:
```

Comment:
`NULLCHECK` does not produce values and gets the "void" VN, so, how does this work at all...?

Reply:
`optIsCSEcandidate()` has 2 purposes.

1. During loop hoisting we check if it will be handled by CSE and if yes, we hoist it out of the loop. During CSE phase, when we actually have to do the expression elimination. By telling that it is CSE candidate, it helps hoisting the check out of loop (because it is an invariant). While working on https://github.com/dotnet/runtime/pull/68061, I realized cases where we would hoist the field address calculation ahead of null-check although there might not be a visible functional difference. But this PR fixes that to some extent at least for nullchecks around `this`.

![image](https://user-images.githubusercontent.com/12488060/165593349-30502e70-a634-46ce-98f6-19f1588cf19f.png)

3. While doing actual CSE, we use `optIsCSEcandidate()`. However, right after that, we check if has any reserved VN and if yes, skip CSEing it.

https://github.com/dotnet/runtime/blob/8f7761af3763d0b84acefbb117ed6becd83061d1/src/coreclr/jit/optcse.cpp#L811-L815

Later, global propagation phase takes care of eliminating the null-check that is present inside the loop.

Reply:
Indeed.

Until now, it has been the case that a "hoistable tree" is (almost, modulo one struct pessimization) the same as "CSEable tree", and both must be values, as the design of hoisting is that we rely on CSE to "clean things up".

But if we start to hoist non-value trees (which is a perfectly valid thing to do), that's now really not related to CSE, because: a) non-value trees will never be actual CSE candidates, b) the "cleanup" phase is assertion propagation.

So that's why I think it is confusing to put this check inside `optIsCSECandidate`.

Reply:
I think based on the above discussion it would be good to add a comment in `IsNodeHoistable` saying that we also can hoist null checks since assertion prop deals with the redundant check left in the loop. It sort of ties loop hoisting together with assertion prop in the same way it is tied together with CSE.

Also, as @SingleAccretion pointed out, I think you can allow this in more places than just for 'this' pointer. We hoist a tree only if its children are invariant which should be safe enough in general for GT_NULLCHECK. Of course that assumes/hopes assertion prop can pick up the additional cases.

Reply:
I tried experimenting with relaxing the hoisting for other object's nullchecks and there are some wins, but I see some regressions of this sort. 

![image](https://user-images.githubusercontent.com/12488060/165810969-686c3ce4-2ebb-4a1f-9fbf-22ae3960889a.png)

Basically, we overcome the number of variables hoisted out of the loop as compared to the available registers. I have plans to revisit these heuristics (specially for arm64) in or after https://github.com/dotnet/runtime/pull/68061 and I would look into "other object's nullcheck hoisting" at that time.

```
      [000167] hoistable    tree cost too low: 3 < 6 (loopVarCount 6 >= availableRegCount 6)
   ... not profitable to hoist
```

> it would be good to add a comment 

If there are no other changes to be done, I would rather do it in my https://github.com/dotnet/runtime/pull/68061

Reply:
Is the new on the right? It looks like an improvement to me, both in size and perf score (the loop body is smaller and with fewer memory accesses). Or am I missing something?

Reply:
True, I was hoping to also get `lea` too out of the loop and it was not (because of reasons I stated above). I reran the `jit-analyze` with PerfScore and I did some improvements. Let me push the changes that enables this for other objects.

Reply:
Makes sense. I think there is still work to be done in hoisting, especially with regards to hoisting more invariant statements.